SharpCaster/ChromeCastClient.cs

```diff
@@ -1,222 +1,272 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Diagnostics;
+﻿using Sharpcaster.Core.Channels;
+using Sharpcaster.Core.Interfaces;
+using Sharpcaster.Core.Models;
+using System;
 using System.IO;
-using System.Linq;
+using System.Net.Security;
+using System.Net.Sockets;
 using System.Threading;
 using System.Threading.Tasks;
 using Extensions.Api.CastChannel;
-using SharpCaster.Channels;
-using SharpCaster.Controllers;
-using SharpCaster.Extensions;
-using SharpCaster.Models;
-using SharpCaster.Models.ChromecastStatus;
-using SharpCaster.Models.MediaStatus;
-using SharpCaster.Interfaces;
-using SharpCaster.Services;
-
-namespace SharpCaster
+using static Extensions.Api.CastChannel.CastMessage.Types;
+using System.Text;
+using Newtonsoft.Json;
+using Sharpcaster.Core.Messages;
+using System.Collections.Generic;
+using System.Linq;
+using Sharpcaster.Extensions;
+using System.Collections.Concurrent;
+using System.Reflection;
+using Sharpcaster.Core.Models.ChromecastStatus;
+using System.Diagnostics;
+using Microsoft.Extensions.DependencyInjection;
+using Sharpcaster.Logging;
+
+namespace Sharpcaster
 {
-    public class ChromeCastClient
+    public class ChromecastClient : IChromecastClient
     {
-        public Volume Volume
+        private const int RECEIVE_TIMEOUT = 30000;
+        private static readonly object LockObject = new object();
+        private TcpClient _client;
+        private Stream _stream;
+        private ChromecastReceiver _receiver;
+        private TaskCompletionSource<bool> ReceiveTcs { get; set; }
+        public static readonly ILog Logger = LogProvider.For<ChromecastClient>();
+        private SemaphoreSlim SendSemaphoreSlim { get; } = new SemaphoreSlim(1, 1);
+
+        private IDictionary<string, Type> MessageTypes { get; set; }
+        private IEnumerable<IChromecastChannel> Channels { get; set; }
+        private ConcurrentDictionary<int, object> WaitingTasks { get; } = new ConcurrentDictionary<int, object>();
+
+        public ChromecastClient()
```

Comment:
I see that you're using `Microsoft.Extensions.DependencyInjection` for dependency injection, nice!

I think it could be improved by not building a "hidden container", but by providing an extension method like in http://asp.net-hacker.rocks/2017/03/06/using-dependency-injection-in-multiple-projects.html

```csharp
public static IServiceCollection AddSharpCaster(this IServiceCollection services)
{
    // add DI setup code here, like: 
    /*
            services.AddTransient<IChromecastChannel, ConnectionChannel>();
            services.AddTransient<IChromecastChannel, HeartbeatChannel>();
            services.AddTransient<IChromecastChannel, ReceiverChannel>();
            services.AddTransient<IChromecastChannel, MediaChannel>();
            var messageInterfaceType = typeof(IMessage);
            foreach (var type in (from t in typeof(IConnectionChannel).GetTypeInfo().Assembly.GetTypes()
                                  where t.GetTypeInfo().IsClass && !t.GetTypeInfo().IsAbstract && messageInterfaceType.IsAssignableFrom(t)
                                  select t))
            {
                services.AddTransient(messageInterfaceType, type);
            }
    */
}
```

Also, by using dependency injection it would not be needed to call the constructor with IServiceCollection, but you can "let the dependency injection do the calling" and inject the interfaces of the needed dependencies in the constructors.

What do you think?